File: OrchidCore/src/main/java/com/eden/orchid/impl/compilers/parsers/TOMLParser.java
package com.eden.orchid.impl.compilers.parsers;

import com.eden.orchid.api.compilers.OrchidParser;
import com.moandjiezana.toml.Toml;
import com.moandjiezana.toml.TomlWriter;

import javax.inject.Inject;
import java.util.Map;
import java.util.regex.Pattern;

public final class TOMLParser extends OrchidParser {

    @Inject
    public TOMLParser() {
        super(100);
    }

    @Override
    public String getDelimiter() {
        return Pattern.quote(getDelimiterString());
    }

    @Override
    public String getDelimiterString() {
        return "+";
    }

    @Override
    public String[] getSourceExtensions() {
        return new String[] {"tml", "toml"};
    }

    @Override
    public Map<String, Object> parse(String extension, String input) {
        Toml toml = new Toml().read(input);
        return toml.toMap();
    }

    @Override
    public String serialize(String extension, Object input) {
        TomlWriter tomlWriter = new TomlWriter();
        return tomlWriter.write(input);
    }
}

File: OrchidCore/src/main/kotlin/com/eden/orchid/impl/compilers/parsers/TOMLParser.kt
package com.eden.orchid.impl.compilers.parsers

import com.eden.orchid.api.compilers.OrchidParser
import com.eden.orchid.utilities.logSyntaxError
import com.moandjiezana.toml.Toml
import com.moandjiezana.toml.TomlWriter
import java.util.regex.Pattern
import javax.inject.Inject

class TOMLParser
@Inject
constructor() : OrchidParser(100) {

    override fun getDelimiter(): String {
        return Pattern.quote(delimiterString)
    }

    override fun getDelimiterString(): String {
        return "+"
    }

    override fun getSourceExtensions(): Array<String> {
        return arrayOf("tml", "toml")
    }

    override fun parse(extension: String, input: String): Map<String, Any>? {
        try {
            val toml = Toml().read(input)
            return toml.toMap()
        }
        catch (e: Exception) {
            val errorMessage = e.message ?: ""

            val lineNumber = "^.*?line (\\d+):.*$"
                .toRegex()
                .matchEntire(errorMessage)
                ?.groupValues
                ?.get(1)
                ?.toIntOrNull()
                ?: 0

            input.logSyntaxError(extension, lineNumber, 0, errorMessage)
        }

        return null
    }

    override fun serialize(extension: String, input: Any?): String {
        val tomlWriter = TomlWriter()
        return tomlWriter.write(input)
    }
}
